Added ROS parameter to toggle performance metrics in gazebo_ros

adityapande-1995 opened the pull request:
Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 committed:
Added ROS parameter to toggle performance metrics

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

jacobperron commented on gazebo_ros/src/gazebo_ros_init.cpp:
```diff
@@ -206,7 +210,12 @@ void GazeboRosInitPrivate::onPerformanceMetrics(
+  // Check if parameter was enabled
+  bool check_enable_performance_metrics ;
+  this->ros_node_->get_parameter("enable_performance_metrics", check_enable_performance_metrics);
+  if (check_enable_performance_metrics){
```
IIUC, we don't need to execute the rest of the logic in `onPerformanceMetrics` above. I suggest we check this parameter at the beginning of the method and return early if the parameter value is `false`.

adityapande-1995 replied:
Done

jacobperron commented on gazebo_ros/src/gazebo_ros_init.cpp:
```diff
@@ -176,6 +176,10 @@ void GazeboRosInit::Load(int argc, char ** argv)
+  // PerformanceMetrics parameter
+  impl_->ros_node_->declare_parameter<bool>("enable_performance_metrics", true);
+
+
```
nitpick: remove this extra blank line (I thought a linter might complain, but perhaps not).

adityapande-1995 replied:
Done

adityapande-1995 committed:
Uncrustify issue fixed

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

jacobperron commented on gazebo_ros/src/gazebo_ros_init.cpp:
```diff
@@ -176,6 +176,10 @@ void GazeboRosInit::Load(int argc, char ** argv)
   impl_->throttler_ = Throttler(rate_param.get<double>());
 
+  // PerformanceMetrics parameter
+  impl_->ros_node_->declare_parameter<bool>("enable_performance_metrics", true);
```
I don't think the performance metrics feature is really documented anywhere. It might be nice to pass a [ParameterDescriptor](https://github.com/ros2/rclcpp/blob/0034929eef9f6626530d8207030b791e121a5415/rclcpp/include/rclcpp/node.hpp#L434) here with the `description` field set to document what this parameter does (e.g. "if set to true, performance metrics are published to the topic 'performance_metrics'").

adityapande-1995 replied:
Done

jacobperron reviewed (commented):
:+1: 

I've left some minor comments

adityapande-1995 committed:
Added quick return logic, parameter description

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

jacobperron commented on gazebo_ros/src/gazebo_ros_init.cpp:
```diff
@@ -188,25 +195,29 @@ void GazeboRosInit::Load(int argc, char ** argv)
+  // Check if performance metrics parameter was enabled
+  bool check_enable_performance_metrics;
+  this->ros_node_->get_parameter("enable_performance_metrics", check_enable_performance_metrics);
+  if (check_enable_performance_metrics) {
```
Prefer returning early, this cuts down on the level of indentation (and is more readable IMO). E.g.

```c++
if (!check_enable_performance_metrics) {
  return;
}

adityapande-1995 replied:
Done

adityapande-1995 committed:
Quick return logic changed for readablity

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

jacobperron reviewed (approved):
LGTM!

It'd be nice to sort out what's up with CI before merging though.

adityapande-1995 committed:
launch test added to disable performace_metrics

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 committed:
Added path

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

jacobperron commented:
@ros-pull-request-builder retest this please

jacobperron commented on gazebo_ros/test/perf_metrics_subscriber.py:
```diff
@@ -0,0 +1,55 @@
+    def read_msgs(self):
+        print("Reading the performance_metrics topic...")
+        os.environ['PYTHONUNBUFFERED'] = '1'
+        self.read_msgs_process = subprocess.Popen(["ros2", "topic", "echo", "/performance_metrics"], stdout=subprocess.PIPE)
```
Rather than depending on ros2topic and starting a subprocess, we can instead create a subscription to /performance_metrics directly in this process. This way we avoid potential bugs (or changes in output) in ros2topic effecting our test or issues with clean shutdown of this test, for example.

jacobperron commented on gazebo_ros/test/perf_metrics_subscriber.py:
```diff
@@ -0,0 +1,55 @@
+
+            now = time.time()
+            if now - start > 5:
+                break
```
I suppose it's a matter of preference, but if we make this the condition of the while-loop it makes it obvious under what circumstances the loop ends, e.g.

```py
while time.time() - start < 5:
    ...
```

adityapande-1995 committed:
Changed test script to use rclpy

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+
+class MyTestingClass(unittest.TestCase):
+    def test_routine(self):
+        rclpy.init()
```
I suggest doing the rclpy.init() and rclpy.shutdown() in [`setUpClass`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUpClass) and [`tearDownClass`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDownClass) methods respectively. This should help remove a lot the if-else blocks below.

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+        node = MakeTestNode()
+        node_found = node.wait_for_gazebo_node()
+
+        if node_found:
```
We could more simply write:

```python
assert node.wait_for_gazebo_node(), 'Parameter could not be set'
```

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+from gazebo_msgs.msg import PerformanceMetrics
+from rclpy.node import Node
+from rcl_interfaces.msg import Parameter, ParameterValue
+from ros2param.api import call_set_parameters
```
If we can help it, I think we should avoid depending on `ros2param` for similar reasons to depending on `ros2topic`. I'd rather just add our own implementation of calling the set_parameters service to this file (e.g. as part of the the test fixture).

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+        if node_found:
+            response = node.set_parameter(False)
+            if response.successful:
+                print("enable_performance_metrics successfully set to false !")
```
Style: prefer single quotes for strings (see [style guide](https://docs.ros.org/en/rolling/Contributing/Code-Style-Language-Versions.html#python)).

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+                else:
+                    print("No messages were published")
+                    rclpy.shutdown()
+                    assert True
```
```suggestion
                assert node.msg_count == 0, f'Received {node.msg_count} messages'
```

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+
+        if node_found:
+            response = node.set_parameter(False)
+            if response.successful:
```
Alternatively,

```python
assert response.successful
```

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+        temp_value = ParameterValue()
+        temp_value.bool_value = state
+        temp_value.type = 1
+        parameter.value = temp_value
```
I think we can avoid the temporary. Also, prefer using the named constants defined in the message instead of literal numbers:
```suggestion
        parameter.value.bool_value = state
        parameter.value.type = ParameterType.PARAMETER_BOOL
```

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+            'performance_metrics',
+            self.subscriber_callback,
+            10)
+        self.subscription
```
```suggestion
```

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+
+
+class MyTestingClass(unittest.TestCase):
+    def test_routine(self):
```
Let's pick a less generic name for our test fixture, like `TestPerformanceMetricsParam`. Same goes for the test method (e.g. `test_disable_performance_metrics` makes it more clear what this test is about). Adding docstrings is a good habit too :slightly_smiling_face:

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+            if 'gazebo' in active_nodes:
+                self.flag_gazebo_node = True
+                print("Found the gazebo node !")
+                time.sleep(0.1)
```
```suggestion
            self.flag_gazebo_node = 'gazebo' in active_nodes
            time.sleep(0.1)
```

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+        self.flag_gazebo_node = False
+        self.msg_count = 0
+
+    def wait_for_gazebo_node(self):
```
It might be nice to pass the timeout as a parameter, e.g.
```suggestion
    def wait_for_gazebo_node(self, timeout_sec=5.0):
```

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+            print("Gazebo node was not found !")
+        return self.flag_gazebo_node
+
+    # Set enable_performance_metrics parameter to false
```
This comment isn't quite correct, since the value depends on the argument to this method. Consider adding a docstring :)

jacobperron commented on gazebo_ros/test/disable_perf_metrics.test.py:
```diff
@@ -0,0 +1,126 @@
+
+class MakeTestNode(Node):
+    def __init__(self):
+        super().__init__('testing_node')
```
Consider a more descriptive name for this test node (e.g. `test_perf_metrics_node`). In case someone notices the node on the ROS graph they'll know roughly where it's coming from.

jacobperron reviewed (commented):
Nice, I like the direction the tests are headed. I think we can do a bit of refactoring to clean it up (see comments below). Also, for completeness we could add a second test method for setting the performance metrics parameter to `true`.

adityapande-1995 committed:
Code cleanup

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 commented:
Changed made:
* Added doc-strings, fixed incorrect comments
* Changed if-else blocks to assertions
* Added timeout parameter to wait_for_gazebo()
* Meaningful / descriptive names for node, file, classes
* Double quotes changed to single quotes, except in doc-strings
* rclpy.init() and shutdown() in ```@classmethod ```
* Shorter logic in wait_for_gazebo() method
* Test case added to enable the parameter and ensure messages are received
*  ```ParameterType.PARAMETER_BOOL``` added instead of hard coding the numerical value
* Unnecessary line ```self.subscription``` removed

TODO:
* Remove dependency on ros2params API
* Linting, uncrustify check

adityapande-1995 committed:
CMake line changed

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 committed:
added set_parameter method based on ros2param call_set_parameter api

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 commented:
set_parameter is based on [ros2param call_set_parameter function](https://github.com/ros2/ros2cli/blob/6cbe07025506233120703bc88e6a2123615e2e97/ros2param/ros2param/api/__init__.py#L189) rather than using the function itself.

adityapande-1995 committed:
pep257 fixes

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 commented:
TODO:
* Reflect the changes made in documentation

jacobperron commented on gazebo_ros/test/test_disable_performance_metrics.test.py:
```diff
@@ -0,0 +1,172 @@
+@pytest.mark.launch_test
+def generate_test_description():
+    """Launch a gzserver node."""
+    os.chdir(os.path.dirname(__file__))
```
I guess this line isn't necessary?

jacobperron commented on gazebo_ros/test/test_disable_performance_metrics.test.py:
```diff
@@ -0,0 +1,172 @@
+    def tearDownClass(cls):
+        rclpy.shutdown()
+
+    def test_parameter_enable(self):
```
For completeness, it would be nice to add a second test for disabling the parameter.

adityapande-1995 replied:
The ``TestPerformanceMetricsParam_disable`` block disables the parameter (when it is originally enabled) and makes sure no msgs are received. After that ``TestPerformanceMetricsParam_enable`` test runs which enables the parameter (which is currently disabled) and checks that messages are received successfully.

jacobperron commented on gazebo_ros/test/test_disable_performance_metrics.test.py:
```diff
@@ -0,0 +1,172 @@
+        assert response.successful, 'Parameter could not be set to True'
+
+        node.start_subscriber()
+        time.sleep(2.5)
+        assert node.msg_count != 0, 'Did not receive any messages after\
+            setting enable_performance_metrics to True, test failed'
```
We can do better than blindly sleeping. We could poll with a timeout for `node.msg_count` to be greater than zero. Or even better, use a Python [`Event`](https://docs.python.org/3/library/threading.html#event-objects) object to signal from the spin thread to this thread.

adityapande-1995 replied:
Added ``Event()`` based quick return logic

jacobperron reviewed (commented):
A few more comments about the tests. Generally looks good to me :slightly_smiling_face:

adityapande-1995 committed:
Removed os.chdir line

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 committed:
Added Event() based quick return logic

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 committed:
Dynamic timeout

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

jacobperron commented on gazebo_ros/test/test_disable_performance_metrics.test.py:
```diff
@@ -0,0 +1,190 @@
+            setting enable_performance_metrics to True, test failed'
+
+
+class TestPerformanceMetricsParam_2(unittest.TestCase):
```
I don't think we need a separate test fixture for the second test; we can keep both tests associated with the same fixture,  ie.

```py
class TestPerformanceMetricsParam(unittest.TestCase):
    
    @classmethod
    def setUpClass(cls):
        rclpy.init()

    @classmethod
    def tearDownClass(cls):
        rclpy.shutdown()

    def test_parameter_enable(self):
        ...
    
    def test_parameter_disable(self):
        ...
```

adityapande-1995 replied:
Done. However, ```rclpy.shutdown()`` has to be called after the end of both of them to stop the spin thread, so that the next node can spin.

adityapande-1995 committed:
Test cases merged to 1 fixture

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 commented:
@ros-simulation/pull-request-builder-bot retest this please

jacobperron commented:
I think CI is blocked by https://github.com/ros/rosdistro/pull/30344

jacobperron reviewed (approved):
Tests look good

adityapande-1995 commented:
@ros-simulation/pull-request-builder-bot retest this please

adityapande-1995 commented:
@ros-simulation/pull-request-builder-bot retest this please

adityapande-1995 committed:
flake8 error, removed unused import

Signed-off-by: Aditya Pande <aditya050995@gmail.com>

adityapande-1995 commented:
The only error in CI currently is ```what():  boost thread: trying joining itself: Resource deadlock avoided``` and it seems flaky.
